feat: add automatic version bumping workflow #1946

error9098x opened the pull request:
<!--- Pull request titles should follow conventional commit format. -->

fixes #1946 

<!--
Thanks for opening a PR! Your contribution is much appreciated.
In order to make sure your PR is handled as smoothly as possible we request that you follow the checklist sections below.
Choose the right checklist for the change that you're making:
-->

## Changes proposed in this pull request

- Added a new workflow file `.github/workflows/version_bump.yml`.
- The workflow uses `pnpm version prerelease` to automatically bump the version.



<!--
Provide a succinct description of what this pull request entails.
-->

Please do let me know if there is any problem with this solution and any other possible improvements.  @mkurapov 

## Context
<!--
What were you trying to do?
Provide further details about how the feature should be tested/reviewed if necessary.
Link issues here -  using `fixes #number`
-->
- The workflow runs on every push to the `main` branch, bumps the prerelease version in `package.json` using the `pnpm version prerelease` command, and pushes the changes back to the `main` branch.

## Checklist
<!--
Checklist items become clickable check boxes once the pull request is created. There is no need to edit them now.
-->

- [ ] Related issues linked using `fixes #number`
- [ ] Tests added/updated
- [ ] Documentation added
- [ ] Make sure that all checks pass
- [ ] Postman collection updated

sabineschaller reviewed (commented):
Hi! Thank you for the contribution! 
We would only like to bump the version when we tag. So let's say we tag main with 'v1.0.0-alpha.4", we would also want that to be reflected in the package.json. Also, when we finally get out of alpha into beta or even into release and let's say we tag with "v1.0.0", that should be the version in the package.json as well.

error9098x commented:
> We would only like to bump the version when we tag. So let's say we tag main with 'v1.0.0-alpha.4", we would also want that to be reflected in the package.json. Also, when we finally get out of alpha into beta or even into release and let's say we tag with "v1.0.0", that should be the version in the package.json as well.

Sure thing, I will add the workflow change the version according to the tag :)

error9098x commented:
> Hi! Thank you for the contribution! We would only like to bump the version when we tag. So let's say we tag main with 'v1.0.0-alpha.4", we would also want that to be reflected in the package.json. Also, when we finally get out of alpha into beta or even into release and let's say we tag with "v1.0.0", that should be the version in the package.json as well.

Hi, I have updated my code accordingly. Now, it checks for new tag and updated the version in package.json, please do tell me if anything else has to be updated.

sabineschaller commented:
Looks good to me but I asked @mkurapov to also have a look since he did most of our GH workflows. Thank you!

mkurapov commented:
@sabineschaller 
I think we should rethink this, since we need the new tag + package version bump _before_ we actually draft & publish a release. What do you think of having a workflow that is just manually ran to create a PR that bumps the package version?

mkurapov commented on .github/workflows/version_bump.yml:
```diff
@@ -0,0 +1,31 @@
+          git config --global user.email 'github-actions@github.com'
+          pnpm version ${GITHUB_REF#refs/tags/v} --no-git-tag-version 
+          git commit -am "Bump version"
+          git push origin main
```
Since `main` is a protected branch, we would need to instead create a PR before merging into `main`.
We could use the https://github.com/marketplace/actions/create-pull-request action here to do the creating/updating of a PR. 

The `pnpm version ${GITHUB_REF#refs/tags/v} --no-git-tag-version` looks good though :+1:

mkurapov commented on .github/workflows/version_bump.yml:
```diff
@@ -0,0 +1,31 @@
+        with:
+          node-version: '18'
+
+      - name: Install dependencies
+        run: pnpm install
```
I think we can get away without installing all of the packages, just need to install `pnpm` using `- uses: pnpm/action-setup@v2`

mkurapov commented on .github/workflows/version_bump.yml:
```diff
@@ -0,0 +1,31 @@
+name: Version Bump
+
+on:
+  create: # It will trigger whenever a new tag or branch is created
```
This will probably be triggered by a manual workflow instead, pending https://github.com/interledger/rafiki/pull/1976#issuecomment-1757881005

sabineschaller commented:
> I think we should rethink this, since we need the new tag + package version bump _before_ we actually draft & publish a release.

Yes, you are right. I was somehow hoping we could do that as a pre-commit hook or something but tagging is not committing...

sabineschaller commented:
@error9098x @mkurapov 
What if we did it the other way around. What if we had a workflow that tagged the repo when we push a change to main that includes a version bump in the package.json?
